tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -692,608 +505,989 @@ void fillShaderBufferFromLongVectorData(std::vector<BYTE> &ShaderBuffer,
+  DISPATCH(TYPE, Initialize<TYPE>, " -DFUNC_INITIALIZE=1")
 
   switch (OpType) {
   case UnaryOpType_Initialize:
-    SpecialDefines = " -DFUNC_INITIALIZE=1";
+    DISPATCH_INITIALIZE(HLSLBool_t);
+    DISPATCH_INITIALIZE(int16_t);
+    DISPATCH_INITIALIZE(int32_t);
+    DISPATCH_INITIALIZE(int64_t);
+    DISPATCH_INITIALIZE(uint16_t);
+    DISPATCH_INITIALIZE(uint32_t);
+    DISPATCH_INITIALIZE(uint64_t);
+    DISPATCH_INITIALIZE(HLSLHalf_t);
+    DISPATCH_INITIALIZE(float);
+    DISPATCH_INITIALIZE(double);
+    break;
+  case UnaryOpType_EnumValueCount:
     break;
-  default:
-    LOG_ERROR_FMT_THROW(L"Unsupported UnaryOpType: %ls", OpTypeName.c_str());
   }
 
-  auto ComputeFunc = [this](const T &A) {
-    return this->computeExpectedValue(A);
-  };
-  InitUnaryOpValueComputer<T>(ComputeFunc);
-}
-
-template <typename T>
-T UnaryOpTestConfig<T>::computeExpectedValue(const T &A) const {
-  if (OpType != UnaryOpType_Initialize) {
-    LOG_ERROR_FMT_THROW(L"computeExpectedValue(const T &A, "
-                        L"UnaryOpType OpType) called on an "
-                        L"unrecognized unary op: %ls",
-                        OpTypeName.c_str());
-    return T();
-  }
+#undef DISPATCH_INITIALIZE
+#undef DISPATCH
 
-  return T(A);
+  LOG_ERROR_FMT_THROW(L"DataType '%s' not supported for UnaryOpType '%s'",
+                      (const wchar_t *)Config.DataType,
+                      (const wchar_t *)Config.OpTypeEnum);
 }
 
-template <typename T>
-UnaryMathOpTestConfig<T>::UnaryMathOpTestConfig(
-    const OpTypeMetaData<UnaryMathOpType> &OpTypeMd)
-    : TestConfig<T>(OpTypeMd), OpType(OpTypeMd.OpType) {
+//
+// UnaryMathOp
+//
+
+template <typename T, typename OUT_TYPE>
+void dispatchUnaryMathOpTest(const TestConfig &Config, UnaryMathOpType OpType,
+                             size_t VectorSize, OUT_TYPE (*Calc)(T)) {
 
-  BasicOpType = BasicOpType_Unary;
+  ValidationConfig ValidationConfig;
 
   if (isFloatingPointType<T>()) {
-    Tolerance = 1;
-    ValidationType = ValidationType_Ulp;
+    ValidationConfig = ValidationConfig::Ulp(1.0);
   }
 
-  switch (OpType) {
-  case UnaryMathOpType_Sign: {
-    // Sign has overridden special logic.
-    auto ComputeFunc = [this](const T &A) { return this->sign(A); };
-    InitUnaryOpValueComputer<int32_t>(ComputeFunc);
-    break;
-  }
-  case UnaryMathOpType_Frexp:
-    // Don't initialize a ValueComputer, Frexp has special logic for handling
-    // its output
-    SpecialDefines = " -DFUNC_FREXP=1";
-    break;
-  default: {
-    auto ComputeFunc = [this](const T &A) {
-      return this->computeExpectedValue(A);
-    };
-    InitUnaryOpValueComputer<T>(ComputeFunc);
+  dispatchUnaryTest(Config, ValidationConfig, OpType, VectorSize, Calc, "");
+}
+
+template <typename T> struct UnaryMathOps {
+  static T Abs(T V) {
+    if constexpr (std::is_unsigned_v<T>)
+      return V;
+    else
+      return static_cast<T>(std::abs(V));
```

Comment:
nit:

```suggestion
    if constexpr (std::is_unsigned_v<T>)
      return V;
    return static_cast<T>(std::abs(V));
```

Reply:
:) that's what I wrote first time round.  Turns out these aren't equivalent.  In the `else` case it'll only expand the if or the else part.  In the early-out case it'll expand both parts - which fails because there is no unsigned version of `std::abs`.